src/backend/distributed/utils/acquire_lock.c

```diff
@@ -0,0 +1,311 @@
+	snprintf(worker.bgw_name, BGW_MAXLEN,
+			 "Citus Lock Acquire Helper: %d/%u",
+			 backendToHelp, MyDatabaseId);
+#if PG_VERSION_NUM >= 110000
+	snprintf(worker.bgw_type, BGW_MAXLEN, "citus_lock_aqcuire");
+#endif
+
+	worker.bgw_flags = BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_start_time = BgWorkerStart_RecoveryFinished;
+	worker.bgw_restart_time = BGW_NEVER_RESTART;
+
+	snprintf(worker.bgw_library_name, BGW_MAXLEN, "citus");
+	snprintf(worker.bgw_function_name, BGW_MAXLEN, "LockAcquireHelperMain");
+	worker.bgw_main_arg = Int32GetDatum(backendToHelp);
+	worker.bgw_notify_pid = 0;
+
+	/*
+	 * we check if args fits in bgw_extra to make sure it is safe to copy the data. Once
+	 * we exceed the size of data to copy this way we need to look into a different way of
+	 * passing the arguments to the worker.
+	 */
+	Assert(sizeof(worker.bgw_extra) >= sizeof(args));
+	memcpy(worker.bgw_extra, &args, sizeof(args));
+
+	if (!RegisterDynamicBackgroundWorker(&worker, &handle))
+	{
+		ereport(ERROR, (errmsg("could not start lock acquiring background worker to "
+							   "force the update"),
+						errhint("Increasing max_worker_processes might help.")));
+	}
+
+
+	MemoryContextCallback *cb = palloc0(sizeof(MemoryContextCallback));
+	cb->func = EnsureStopLockAcquireHelper;
+	cb->arg = handle;
+
+	MemoryContextRegisterResetCallback(CurrentMemoryContext, cb);
+
+	return handle;
+}
+
+
+/*
+ * EnsureStopLockAcquireHelper is designed to be called as a MemoryContextCallback. It
+ * takes a handle to the background worker and Terminates it. It is safe to be called on a
+ * handle that has already been terminated due to the guard around the generation number
+ * implemented in the handle by postgres.
+ */
+static void
+EnsureStopLockAcquireHelper(void *arg)
+{
+	BackgroundWorkerHandle *handle = (BackgroundWorkerHandle *) arg;
+	TerminateBackgroundWorker(handle);
+}
+
+
+/*
+ * Signal handler for SIGTERM
+ *		Set a flag to let the main loop to terminate, and set our latch to wake
+ *		it up.
+ */
+static void
+lock_acquire_helper_sigterm(SIGNAL_ARGS)
+{
+	int save_errno = errno;
+
+	got_sigterm = true;
+	SetLatch(MyLatch);
+
+	errno = save_errno;
+}
+
+
+/*
+ * ShouldAcquireLock tests if our backend should still proceed with acquiring the lock,
+ * and thus keep terminating conflicting backends. This function returns true until a
+ * SIGTERM, background worker termination signal, has been received.
+ *
+ * The function blocks for at most sleepms when called. During operation without being
+ * terminated this is the time between invocations to the backend termination logic.
+ */
+static bool
+ShouldAcquireLock(long sleepms)
+{
+	int rc;
+
+	rc = WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
+				   sleepms * 1L, PG_WAIT_EXTENSION);
+	ResetLatch(MyLatch);
+
+	/* emergency bailout if postmaster has died */
+	if (rc & WL_POSTMASTER_DEATH)
+	{
+		proc_exit(1);
+	}
+
+	CHECK_FOR_INTERRUPTS();
```

Comment:
I wonder if we should just do a graceful proc_exit here instead of throwing an error.

Reply:
I am not completely sure what happens in `CHECK_FOR_INTERRUPTS` that would cause the ungraceful `proc_exit`. This logic is extracted from an example in the postgres source code of [`src/test/modules/worker_spi`](https://github.com/postgres/postgres/blob/414cca40d506dd3f17b49ae3139853139192c2ba/src/test/modules/worker_spi/worker_spi.c#L228-L243).

This code is based on that as I assumed that is best practice on the signal handling in a worker (albeit I skipped sighup for now as we don't use any guc's).
Let me know if you have reasons to believe the `CHECK_FOR_INTERRUPTS` here would cause an issue.

Reply:
CHECK_FOR_INTERRUPT() throws an error here: https://github.com/postgres/postgres/blob/master/src/backend/tcop/postgres.c#L3137 .  This is normally used to interrupt queries and abort the transaction.

If an error is thrown in a background worker and not caught anywhere then that is seen as a failure and the worker exits with error code 1 and is restarted by postmaster, though that can be prevented by `BGW_NEVER_RESTART`. 

However, in this case getting a cancellation is a perfectly normal, expected outcome, so we might as well exit gracefully.

Reply:
During normal operation we call `TerminateBackgroundWorker` which asks postmaster to send the background worker a `SIGTERM`. This signal is trapped in via `lock_acquire_helper_sigterm` per [`pqsignal`](https://github.com/citusdata/citus/pull/2773/files#diff-2e185cd6be0ac05904cbedc2db8803d1R193).

any other signal is unexpected behaviour I guess and suffices a non clean exit.

I am not sure about the semantics of `pqsignal` with `CHECK_FOR_INTERRUPTS` though, are you saying the latter will handle `SIGTERM` differently in the case where `MyLatch` has stopped due to a timeout and a `SIGTERM` has been been received between line 162 and the `CHECK_FOR_INTERRUPTS`?

During development I have not witnessed any non graceful shutdown of the background worker. For safety reasons the background worker is actually started with `BGW_NEVER_RESTART` as a defensive way of running this background worker.